packages/contour: Expose loadBalancerIP in config

doddatpivotal opened the pull request:
If applied, this commit will all a user to specify desired load balancer IP address for Contour's Envoy service.

Fixes #2622

Signed-off-by: Dodd Pfeffer <dpfeffer@vmware.com>

## Describe testing done for PR
I'm not quite sure how to test this.

## Special notes for your reviewer
I attempted to find all relavent docs that would need to be updated as well.
I felt making the update to the latest patch release of the last two minors was sufficient.  I'm open to feedback.

Commit:
packages/contour: Expose loadBalancerIP in config

If applied, this commit will all a user to specify desired load balancer IP address for Contour's Envoy service.

Fixes #2622

Signed-off-by: Dodd Pfeffer <dpfeffer@vmware.com>

skriss reviewed (changes requested):
Thanks for the PR @doddatpivotal! Looks pretty good, really just one substantive comment on the overlay logic.

> I felt making the update to the latest patch release of the last two minors was sufficient. I'm open to feedback.

Yep that makes sense to me too, we should drop some of the older versions soon anyway.

The last thing we'll need to do is push the updated packages to the registry, and update the `package.yaml` files to have the right SHAs. @seemiller is there a suggested process here for when the contributor is not a package owner / does not have access to the registry? Probably just makes sense for me to do a follow-up PR after this merges to push the packages and update the SHAs?

sunjayBhatia reviewed (commented):
cant think of anything besides @skriss' review 👍🏽

Commit:
packages/contour: Updates exposure of loadBalancerIP

If applied, this commit will remove the undesired default value and guard against lack of existance
of the loadBalancerIP in the upstream manifest.

Fixes #2622

Signed-off-by: Dodd Pfeffer <dpfeffer@vmare.com>

doddatpivotal commented on addons/packages/contour/1.18.2/bundle/config/overlays/update-envoy-service.yaml:
```diff
@@ -25,6 +25,10 @@ spec:
     #@ end
   #@ end
 
+  #@ if data.values.envoy.service.type == "LoadBalancer":
```
Great feedback.  I've updated in a second commit.

skriss replied:
Note the second line in the suggestion too - `#@overlay/match missing_ok=True`, needs to be added since this field is not in the upstream manifest.

doddatpivotal commented on addons/packages/contour/1.18.2/package.yaml:
```diff
@@ -69,6 +69,10 @@ spec:
+                loadBalancerIP:
+                  type: string
+                  description: The desired load balancer IP for the Envoy service.
+                  default: Local
```
Great feedback.  I've updated in a second commit.

doddatpivotal commented:
@skriss I believe I have addressed the substantive feedback.  Please let me know if there is more I can do.

skriss commented:
FYI my basic approach to validating this is to manually edit the `values.yaml` to include the different permutations of envoy service type + load balancer IP, and then to run `ytt -f addons/packages/contour/1.19.1/bundle/config | grep loadBalancerIP` and inspect the output.

Commit:
packages/contour: Updated exposure loadBalancerIP

If applied, this commit will correct issue in ytt template addressing case of upstream manifest not having loadBalancerIP.

Fixes #2622

Signed-off-by: Dodd Pfeffer <dpfeffer@vmware.com>

doddatpivotal commented:
@skriss thank you for the unit test suggestion.  That helped to quickly identify the flaw in my previous commit and have confidence that the new one works.

skriss reviewed (approved):
LGTM, thanks @doddatpivotal! I'll plan to follow up with a PR that pushes the packages to the registry and bumps the package.yaml SHAs.

davidvonthenen reviewed (approved):
deferring review to others in thread here
